text/0000-node-reactions.md

```diff
@@ -0,0 +1,208 @@
+Element.outerHTML
+
+Node.appendChild
+Node.insertBefore
+Node.nodeValue
+Node.replaceChild
+Node.removeChild
+Node.textContent
+
+Range.insertNode
+Range.deleteContents
+Range.extractContents
+Range.surroundContents
+```
+
+#### Reactions for various APIs
+Here are some sample reactions for the APIs listed above.
+- Node.appendChild: 
+    - If the childNode has an existing parent and the parent is connected, invoke disconnectedCallback for the childNode and its subtree.
+    - If new parent is connected, invoke connectedCallback for the childNode and its subtree
+
+- Node.insertBefore: Same steps as Node.appendChild
+
+- Node.replaceChild
+    - if the parent node is connected
+        - invoke disconnectedCallback for the existing child and its subtree
+        - invoke disconnectedCallback for the new child and its subtree
+
+# Drawbacks
+
+Why should we *not* do this? Please consider:
+
+- Performance: Traversing the dom is not cheap. We need to think about memoization techniques to prevent repetitive traversal. This is of significant importance since the invocations are synchronous in nature.
+- Security: Can a malicious actor sniff and get access to hooks that they do not own? If they did, could utilize that hook and attack the subscriber by simulating an event.
+- Native web components: When we switch to native web components, will this library become obsolete?
+
+# Alternatives
+
+## MutationObserver
+ MutationObserver is a browser API to monitor mutations in a node and its subtree. MutationObserver is not a viable option because of the following limitations:
+ 1. Callbacks are invoked asynchronously. This is a major limitation because lifecycle callbacks are to be invoked synchronously as per spec. Asynchronous reactions are also a limitation for applying styling to shadow dom in case of portal elements(TODO: clarify).
+
+# Adoption strategy
+
+The usage of this library in LWC will be an implementation detail of the engine. Component developers should notive no difference in usage.
+
+How should this feature be taught to existing Lightning Web Components developers?
+
+# Unresolved questions
+
+- How will traversal work for native custom elements running in closed shadow mode? In web components, Lifecycle hooks behave the same regardless of mode of shadow DOM. For example: https://jsbin.com/misofod/edit?html,js,console,output
+- Are there any browser APIs that are faster than manual DOM traversal?
+
+Notes:
+1. isConnected works for regular dom and shadow DOM: https://jsbin.com/femohaqoku/edit?html,js,console,output
+2. isConnected works for shadowRoot node and other documentFragments
+3. When a node is moved to a different subtree, invoke its disconnectedCallback first and then the connectedCallback: https://jsbin.com/qawacil/edit?html,js,output
+4. All hooks for a given node is processed first before moving down the subtree: https://jsbin.com/qawacil/edit?html,js,output
+5. Hooks in subtree are invoked even in closed shadow mode.
```

Comment:
6. When a documentFragment is appended to a node, only the children of the documentFragment are connected to the document. So should the connectedCallback of the documentFragment be invoked?